src/tools/illink/src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -705,10 +698,23 @@ void ProcessVirtualMethod (MethodDefinition method)
 		{
 			Annotations.EnqueueVirtualMethod (method);
 
-			var defaultImplementations = Annotations.GetDefaultInterfaceImplementations (method);
-			if (defaultImplementations != null) {
-				foreach (var defaultImplementationInfo in defaultImplementations) {
-					ProcessDefaultImplementation (defaultImplementationInfo.InstanceType, defaultImplementationInfo.ProvidingInterface);
+			if (method.DeclaringType.IsInterface) {
+				var defaultImplementations = Annotations.GetDefaultInterfaceImplementations (method);
+				if (defaultImplementations is not null) {
+					foreach (var dimInfo in defaultImplementations) {
+						ProcessDefaultImplementation (dimInfo.ImplementingType, dimInfo.InterfaceImpl, dimInfo.DefaultInterfaceMethod);
+
+						var ov = new OverrideInformation (method, dimInfo.DefaultInterfaceMethod, Context);
+						if (IsInterfaceImplementationMethodNeededByTypeDueToInterface (ov, dimInfo.ImplementingType))
+							MarkMethod (ov.Override, new DependencyInfo (DependencyKind.Override, ov.Base), ScopeStack.CurrentScope.Origin);
+					}
+				}
+				var overridingMethods = Annotations.GetOverrides (method);
+				if (overridingMethods is not null) {
+					foreach (var ov in overridingMethods ?? []) {
+						if (IsInterfaceImplementationMethodNeededByTypeDueToInterface (ov, ov.Override.DeclaringType))
```

Comment:
Shouldn't that situation be caught above when we loop over the default implementations of each interface method?

Reply:
I thought about this a little more, and it's another one of those "It works all the time, but not because the logic is correct". The DIM logic above handles static interface methods since they cannot be provided by a base type, only a DIM, but If a base method provides an instance method, we'll pass the wrong type in, but if the type implementing the interface is instantiated, the type providing the method for the interface will also be marked as instantiated.

Reply:
Looks like this fix is going to be more involved. I'll make an issue and make a separate PR.

Reply:
Uh oh, I really thought static interface methods couldn't be provided by the base, and I think that assumption might be made in the linker. I'll set up some test cases and see what's broken.

Reply:
Ah, the compiler injects a static interface method on D that call's C's method. [Sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIBMAjALABQ5AEgKYoAO1ATgHQAiA9gDysB8AFAEoA3OXKEABDXpNxAb3LjF4vAGZlRAGzKALOI6cAKvwHiA7gAsm1cQfEhxASTkKlrg8wCygkWVcBfcgCKMlgEJgAzCABjayd5XyU8TXEIYABnBEZohB1xL2FA0TIJAGFnBMVVdS08XXzy11ckgE4+ACIaAE82gorxIKCxAj07cRK4RzkgoA===:~:text=.method%20private%20hidebysig%20static%20%0A%20%20%20%20%20%20%20%20void%20I.M%20()%20cil%20managed%20%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20.override%20method%20void%20I%3A%3AM()%0A%20%20%20%20%20%20%20%20//%20Method%20begins%20at%20RVA%200x2087%0A%20%20%20%20%20%20%20%20//%20Code%20size%206%20(0x6)%0A%20%20%20%20%20%20%20%20.maxstack%208%0A%0A%20%20%20%20%20%20%20%20IL_0000%3A%20call%20void%20C%3A%3AM()%0A%20%20%20%20%20%20%20%20IL_0005%3A%20ret%0A%20%20%20%20%7D%20//%20end%20of%20method%20D%3A%3AI.M)